nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -291,30 +347,197 @@ impl DataStore {
+                    //
+                    // We apply rules before preferences to ensure they are always respected.
+                    // Furthermore, the evaluation of preferences is a target-seeking operation,
+                    // which identifies the distinct sets of targets, and searches them in
+                    // decreasing preference order.
+                    //
+                    // # Logic
+                    //
+                    // ## Background: Notation
+                    //
+                    // We use the following symbols for sets below:
+                    // - ∩: Intersection of two sets (A ∩ B is "everything that exists in A and
+                    // also exists in B").
+                    // - \: difference of two sets (A \ B is "everything that exists in A that does
+                    // not exist in B).
+                    //
+                    // We also use the following notation for brevity:
+                    // - AA,P=Fail: All sleds sharing an anti-affinity instance within a group with
+                    // policy = 'fail'.
+                    // - AA,P=Allow: Same as above, but with policy = 'allow'.
+                    // - A,P=Fail: All sleds sharing an affinity instance within a group with
+                    // policy = 'fail'.
+                    // - A,P=Allow: Same as above, but with policy = 'allow'.
+                    //
+                    // ## Affinity: Apply Rules
+                    //
+                    // - Targets := All viable sleds for instance placement
+                    // - Banned := AA,P=Fail
+                    // - Required := A,P=Fail
+                    // - if Required.len() > 1: Fail (too many constraints).
+                    // - if Required.len() == 1...
+                    //   - ... if the entry exists in the "Banned" set: Fail
+                    //     (contradicting constraints 'Banned' + 'Required')
+                    //   - ... if the entry does not exist in "Targets": Fail
+                    //     ('Required' constraint not satisfiable)
+                    //   - ... if the entry does not exist in "Banned": Use it.
+                    //
+                    // If we have not yet picked a target, we can filter the
+                    // set of targets to ignore "banned" sleds, and then apply
+                    // preferences.
+                    //
+                    // - Targets := Targets \ Banned
+                    //
+                    // ## Affinity: Apply Preferences
+                    //
+                    // - Preferred := Targets ∩ A,P=Allow
+                    // - Unpreferred := Targets ∩ AA,P=Allow
+                    // - Neither := Preferred ∩ Unpreferred
+                    // - Preferred := Preferred \ Neither
+                    // - Unpreferred := Unpreferred \ Neither
+                    // - If Preferred isn't empty, pick a target from it.
```

Comment:
We might want to document this decision externally when we describe how the "allow" policy works, since this means that affinity preferences are "stronger" than anti-affinity preferences when they conflict.

Reply:
I don't *think* it's true that affinity is stronger than anti-affinity. With this ordering, if a target is both "preferred" and "unpreferred", it is not considered in this first batch.

The order is:

- Sleds in preferred and not in unpreferred
- (Sleds that aren't in preferred, and are also not in unpreferred) OR (Sleds that are in both preferred and unpreferred)
- Sleds that are in unpreferred

Another framing:

The priority is "favored" / "neutral" / "unfavored". Belonging to "Just" affinity groups makes the sled "Favored". Belonging to "just" anti-affinity groups makes the sled "unfavored". Belonging to *neither* or *both* causes a sled to be "neutral".

Reply:
I have to confess that I'm unsure what I originally had in mind here, but I agree with your explanation (particularly your reframing: sleds with friendly instances on them get a priority boost; sleds with unfriendly instances get a priority penalty; boosts and penalties cancel each other out; we pick from the highest-priority pool that has at least one candidate sled in it).